plugin.js

```diff
@@ -27,15 +29,33 @@ function factory (options) {
     }
 
     const key = header.substring(bearerType.length).trim()
-    if (authenticate(keys, key) === undefined) {
-      const invalidKeyError = Error('invalid authorization header')
-      fastifyReq.log.error('invalid authorization header: `%s`', header)
-      if (contentType) fastifyRes.header('content-type', contentType)
-      fastifyRes.code(401).send(errorResponse(invalidKeyError))
-      return
+    let retVal
+    if (auth && auth instanceof Function) {
+      try {
+        retVal = auth(key)
+      } catch (err) {
+        retVal = Promise.reject(err)
+      }
+    } else {
+      retVal = authenticate(keys, key)
     }
 
-    next()
+    const invalidKeyError = Error('invalid authorization header')
+    Promise.resolve(retVal).then((val) => {
```

Comment:
Can you please add a comment explaining why we are calling  `Promise.resolve()` here?

Reply:
This is not adding unuseful overhead only to avoid to check if the retVal is or is not a promise?

Reply:
I am not an expert. I made a test with node 10 which I think tests if it adds overhead

p-resolve.js
```js
let i=0; 

function prn(){
  let val = Math.random()*100
  Promise.resolve(val).then((v)=>{
    console.log(val)
  })
  if(i<10000000){
    i++;
    setImmediate(prn);
  }
}
prn();
```
check-then.js
```js
let i=0;
function prn(){
  let val = Math.random()*100
  if(val.then && val.then instanceof Function)
  {
    val.then((v)=>{
      console.log((v));
    })
  }
  else {
    console.log(val);
  }

  if(i<10000000){
    i++;
    setImmediate(prn);
  }
}
prn();
```

Ran them  time node  script.js > /dev/null

promise-resolve: 1m27.168s
check-then: 1m26.207s

I don't know if it is a good test.

Reply:
My thought was about [some performance tips for node <= 8](https://v8.dev/blog/fast-async). For sure the Promise performance has been increased with the latest versions and maybe this check doesn't worth the effort.

A sync function that became async (with promise) add an overhead (in terms of memory, gc execution also, not only throughput). This overhead is of X% (I didn't find the article with benchmark).

That example, in node 8, produce these result on my laptop:
// 157.415 ms if
// 190.002 ms promise

Anyway, we can focus on future versions of node 👍